hilti/include/rt/type-info.h

```diff
@@ -0,0 +1,1071 @@
+// Copyright (c) 2020 by the Zeek Project. See LICENSE for details.
+
+#pragma once
+
+#include <optional>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include <hilti/rt/exception.h>
+#include <hilti/rt/fmt.h>
+#include <hilti/rt/types/all.h>
+#include <hilti/rt/types/bytes.h>
+#include <hilti/rt/types/stream.h>
+
+namespace hilti::rt {
+
+struct TypeInfo;
+
+namespace type_info {
+
+/** Exception flagging incorrect use of type-info values. */
+HILTI_EXCEPTION(InvalidValue, RuntimeError);
+
+class Value;
+
+namespace value {
+
+/**
+ * Helper class to provide safe traversal of HILTI values through the
+ * type-info API. To initiate traversal, an instance of `Parent` is needed
+ * that has its life-time tied to a strong reference encapsulating the value.
+ * The instance will keep the value alive during its own lifetime, and the
+ * traversal will catch if that ends prematurely.
+ */
+class Parent {
+public:
+    /** Constructor that ties existing HILTI value to instance. */
+    template<typename T>
+    Parent(const StrongReference<T>& value) : _handle(std::make_shared<bool>()), _value(value) {}
+
+    /** Constructor that leaves instance initially untied. */
+    Parent() : _handle(std::make_shared<bool>()) {}
+
+    /** Tie instances to an existing HILTI value. */
+    void tie(hilti::rt::StrongReferenceGeneric value) { _value = std::move(value); }
+
+private:
+    friend class type_info::Value;
+
+    std::weak_ptr<bool> handle() const {
+        if ( ! _value )
+            throw InvalidValue("type-info traversal not tied to value");
+
+        return _handle;
+    }
+
+    std::shared_ptr<bool> _handle;
+    std::optional<hilti::rt::StrongReferenceGeneric> _value;
+};
```

Comment:
This helper class gets the job done, but feels a little heavy for me. WDYT about instead of inheriting `ParsedUnit` from `Parent` to just use a `shared_ptr<ParsedUnit*>` when constructing a `Value` (assertions are that the smart pointer contains a value and that the contained value is not nil)?

That would avoid `friend`ing `Value`, the additional inheritance level and from having to distinguish between the different states of `_handle` and `_value`. The downside would be that `ParsedUnit` would have to make a slightly nastier call passing in `make_shared(this)` and would need to check that `value` is not called before `initialize` (maybe just using one of its other members as a proxy, e.g., asserting `_unit`.

Reply:
Problem with that: it'd be a layering violation. The type-info API is on the HILTI side, but ParsedUnit is on the Spicy side, HILTI doesn't know about that. That's why I did it this way. It's also used inside a HILTI-side test, that's what the 1st constructor is for. Let me know if you have another idea, though.